db/version_set.cc

```diff
@@ -625,38 +696,110 @@ void LevelIterator::Prev() {
 }
 
 void LevelIterator::SkipEmptyFileForward() {
-  while (file_iter_.iter() == nullptr ||
-         (!file_iter_.Valid() && file_iter_.status().ok() &&
-          !file_iter_.iter()->IsOutOfBound())) {
-    // Move to next file
-    if (file_index_ >= flevel_->num_files - 1) {
-      // Already at the last file
-      SetFileIterator(nullptr);
+  for (;;) {
+    while (file_iter_.iter() == nullptr ||
+           (!file_iter_.Valid() && file_iter_.status().ok() &&
+            !file_iter_.iter()->IsOutOfBound())) {
+      // Move to next file
+      if (file_index_ >= flevel_->num_files - 1) {
+        // Already at the last file
+        SetFileIterator(nullptr);
+        return;
+      }
+      if (KeyReachedUpperBound(file_smallest_key(file_index_ + 1))) {
+        SetFileIterator(nullptr);
+        return;
+      }
+      InitFileIteratorForward(file_index_ + 1);
+    }
+
+    if (!file_iter_.Valid()) {
       return;
     }
-    if (KeyReachedUpperBound(file_smallest_key(file_index_ + 1))) {
+
+    if (range_tombstone_.end_key_.empty()) {
+      // The range tombstone extends to the end of the sstable.
+      if (range_tombstone_.seq_ == 0) {
+        // The range tombstone doesn't apply to the keys in the
+        // sstable. Return the entry.
+        return;
+      }
+      // The range tombstone is a deletion tombstone. Loop to allow moving to
+      // the next file.
       SetFileIterator(nullptr);
+      continue;
+    }
+
+    auto key = ExtractUserKey(file_iter_.key());
+    if (icomparator_.user_comparator()->Compare(
+            key, range_tombstone_.end_key_) >= 0) {
+      // The key is past the tombstone. Grab the tombstone covering the
+      // current key. Loop so that we can have the proper check for the
+      // tombstone end key.
+      InitRangeTombstone(key);
+      continue;
+    }
+    // The key is contained within the current tombstone.
+    if (range_tombstone_.seq_ == 0) {
+      // The tombstone doesn't apply to the sstable. Return the entry.
       return;
     }
-    InitFileIterator(file_index_ + 1);
-    if (file_iter_.iter() != nullptr) {
-      file_iter_.SeekToFirst();
+    // Find the next non-deleted key.
+    file_iter_.Seek(tombstone_internal_end_key());
```

Comment:
I think this might jump one too far in case there's a Put at the tombstone's end key with newer seqnum. That's because our internal key comparator orders by descending seqnum. We could use `kMaxSequenceNumber` in `tombstone_internal_end_key` rather than the actual seqnum.

Reply:
Yeah, I think you're correct. Your suggestion seems right.